cli/pkg/kctrl/cmd/package/release/helm_openapi_schema_gen.go

```diff
@@ -0,0 +1,373 @@
+	yaml2 "gopkg.in/yaml.v2"
+	yaml3 "gopkg.in/yaml.v3"
+	"k8s.io/apimachinery/pkg/runtime"
+	"sigs.k8s.io/yaml"
+)
+
+// keys used when generating an OpenAPI Document
+const (
+	typeProp        = "type"
+	formatProp      = "format"
+	descriptionProp = "description"
+	itemsProp       = "items"
+	propertiesProp  = "properties"
+	defaultProp     = "default"
+)
+
+var propOrder = map[string]int{
+	typeProp:        1,
+	formatProp:      2,
+	descriptionProp: 3,
+	itemsProp:       4,
+	propertiesProp:  5,
+	defaultProp:     6,
+}
+
+const (
+	objectVal = "object"
+	arrayVal  = "array"
+	floatVal  = "float"
+)
+
+// Yaml tags that differentiate the type of scalar object in the node
+const (
+	nullTag  = "!!null"
+	boolTag  = "!!bool"
+	intTag   = "!!int"
+	floatTag = "!!float"
+)
+
+type MapItem struct {
+	Key   interface{}
+	Value interface{}
+}
+
+type Map struct {
+	Items []*MapItem
+}
+
+func (m *Map) Iterate(iterFunc func(k, v interface{})) {
+	for _, item := range m.Items {
+		iterFunc(item.Key, item.Value)
+	}
+}
+
+type openAPIKeys []*MapItem
+
+func (o openAPIKeys) Len() int {
+	return len(o)
+}
+
+func (o openAPIKeys) Less(i, j int) bool {
+	return propOrder[o[i].Key.(string)] < propOrder[o[j].Key.(string)]
+}
+
+func (o openAPIKeys) Swap(i, j int) {
+	o[i], o[j] = o[j], o[i]
+}
+
+type HelmValuesSchemaGen struct {
+	// dirPath to the helm chart directory
+	dirPath string
+}
+
+func NewHelmValuesSchemaGen(path string) HelmValuesSchemaGen {
+	return HelmValuesSchemaGen{path}
+}
+
+func (h HelmValuesSchemaGen) Schema() (*kcdatav1alpha1.ValuesSchema, error) {
+	fileData, err := h.readValuesFile()
+	if err != nil {
+		return nil, err
+	}
+	if len(fileData) == 0 {
+		return &kcdatav1alpha1.ValuesSchema{
+			OpenAPIv3: runtime.RawExtension{Raw: nil},
+		}, nil
+	}
+
+	var document yaml3.Node
+	err = yaml3.Unmarshal(fileData, &document)
+	if err != nil {
+		return nil, err
+	}
+	if document.Kind == 0 {
+		return &kcdatav1alpha1.ValuesSchema{
+			OpenAPIv3: runtime.RawExtension{Raw: nil},
+		}, nil
+	}
+	if document.Kind != yaml3.DocumentNode {
+		// return proper error message
+		return nil, fmt.Errorf("invalid node kind supplied: %d", document.Kind)
+	}
+	if document.Content[0].Kind != yaml3.MappingNode {
+		return nil, fmt.Errorf("values file must resolve to a map (was %d)", document.Content[0].Kind)
+	}
+	openAPIProperties, err := h.calculateProperties(nil, document.Content[0])
+	if err != nil {
+		return nil, err
+	}
+
+	bytes, err := yaml2.Marshal(h.convertToYAML(openAPIProperties))
+	if err != nil {
+		return nil, err
+	}
+	jsonEncodedBytes, err := yaml.YAMLToJSON(bytes)
+	if err != nil {
+		return nil, err
+	}
+	return &kcdatav1alpha1.ValuesSchema{
+		OpenAPIv3: runtime.RawExtension{Raw: jsonEncodedBytes},
+	}, nil
+}
+
+func (h HelmValuesSchemaGen) readValuesFile() ([]byte, error) {
+	helmChartDir := h.dirPath
+	fileInfo, err := os.Stat(helmChartDir)
+	if err != nil {
+		return nil, fmt.Errorf("unable to read %s: %s", h.dirPath, err.Error())
```

Comment:
os.Stat() always returns file path, so I think we can exclude it while wrapping the error. Also, `unable to read` might not be the best context for this error (example, it could be a not found error).
```suggestion
		return nil, fmt.Errorf("stat: %w", err.Error())
```

Reply:
Seems like os.stat error is pretty self explanatory. We need not wrap it or do anything. 
Sample error:
`stat /abc: no such file or directory`

Reply:
Adopted, just returning the same error now.